pkg/engine/lifecycletest/resource_hooks_test.go

```diff
@@ -0,0 +1,538 @@
+		require.NoError(t, err)
+
+		shouldBeCalled := func(ctx context.Context, urn resource.URN, id resource.ID,
+			newInputs, oldInputs, newOutputs, oldOutputs resource.PropertyMap,
+		) error {
+			hookCalled = true
+			require.Equal(t, urn, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"))
+			expectedNewInputs := map[string]any{
+				"foo":  "bar",
+				"frob": "updated",
+			}
+			require.Equal(t, expectedNewInputs, newInputs.Mappable(), "Hook receieves the new inputs")
+			expectedOldInputs := map[string]any{
+				"foo":  "bar",
+				"frob": "baz",
+			}
+			require.Equal(t, expectedOldInputs, oldInputs.Mappable(), "Hook receieves the old inputs")
+			expectedOldOutputs := map[string]any{
+				"foo":  "bar",
+				"frob": "baz",
+				"baz":  float64(24),
+			}
+			require.Equal(t, expectedOldOutputs, oldOutputs.Mappable(), "Hook receieves the old outputs")
+			require.Nil(t, newOutputs, "there are no new outputs for before update hooks")
+			return nil
+		}
+		shouldBeCalledHook, err := deploytest.NewHook(monitor, callbacks, "shouldBeCalled", shouldBeCalled, true)
+		require.NoError(t, err)
+
+		// On the first run through the program, we'll register `shouldNotBeCalledHook` as a BeforeUpdate hook
+		// for the resource.
+		hooks := []*deploytest.ResourceHook{shouldNotBeCalledHook}
+		if isUpdate {
+			// On the second run through, we switch to another hook. We expect this hook to be called during
+			// the update.
+			hooks = []*deploytest.ResourceHook{shouldBeCalledHook}
+		}
+
+		_, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs: inputs,
+			ResourceHookBindings: deploytest.ResourceHookBindings{
+				BeforeUpdate: hooks,
+			},
+		})
+		require.NoError(t, err)
+
+		err = monitor.SignalAndWaitForShutdown(context.Background())
+		require.NoError(t, err)
+		return nil
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	project := p.GetProject()
+
+	// Run an update to create the resource
+	snap, err := lt.TestOp(Update).RunStep(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil, "0")
+	require.NoError(t, err)
+	require.NotNil(t, snap)
+	require.Len(t, snap.Resources, 2)
+	require.Equal(t, snap.Resources[0].URN.Name(), "default")
+	require.Equal(t, snap.Resources[1].URN.Name(), "resA")
+	require.False(t, hookCalled)
+
+	// change the inputs
+	inputs = resource.NewPropertyMapFromMap(map[string]interface{}{
+		"foo":  "bar",
+		"frob": "updated",
+	})
+	// and use the new hook
+	isUpdate = true
+	snap, err = lt.TestOp(Update).RunStep(project, p.GetTarget(t, snap), p.Options, false, p.BackendClient, nil, "1")
+	require.NoError(t, err)
+	require.NotNil(t, snap)
+	require.Len(t, snap.Resources, 2)
+	require.Equal(t, snap.Resources[0].URN.Name(), "default")
+	require.Equal(t, snap.Resources[1].URN.Name(), "resA")
+	require.True(t, hookCalled)
+}
```

Comment:
Not sure if we're adding in this PR or later but we need a test that a transform can add/remove hooks